tutorial/pytorch_agent/train_cpu_model-pytorch.py

```diff
@@ -110,96 +137,218 @@ def main():
+
+def model_to_script(model, output):
+    model.eval()
+    model_scripted = torch.jit.script(model)
+    model_scripted.save(output)
+    print("Model saved to {}".format(output))
+
+def training_loop(config, input_size, train_set, val_set):
+    train_loader = torch.utils.data.DataLoader(dataset = train_set, batch_size = config['batch_size'], shuffle = False)
+    val_loader = torch.utils.data.DataLoader(dataset = val_set, batch_size = config['batch_size'], shuffle = False)
+
+    model = P3Net(width_input=input_size, width_fc=config["width_fc"], depth_fc=config["depth_fc"])
+    learning_rate = config["lr"]
+    optimizer = torch.optim.Adam(model.parameters(), lr=learning_rate)
+
+    for epoch in range(EPOCH_SIZE):
+        loss, accuracy = train(model, optimizer, train_loader, val_loader)
+
+        with tune.checkpoint_dir(epoch) as checkpoint_dir:
+            path = os.path.join(checkpoint_dir, "checkpoint")
+            torch.save((model.state_dict(), optimizer.state_dict()), path)
+
+        # Send the current training result back to Tune
+        tune.report(loss=loss, accuracy=accuracy)
+
+def train(model, optimizer, train_loader, val_loader):
+    loss_fn = nn.MSELoss()
+
+    for idx, (inputs, target_control) in enumerate(train_loader):
+        model.train()
+        # Clear gradient
+        optimizer.zero_grad()
+        # Get model output
+        predicted_control = model(inputs)
+        # loss calculation vs target
+        loss = loss_fn(predicted_control, target_control)
+        loss.backward()
+        # update model weights
+        optimizer.step()
+
+    model.eval()
+    prediction_total = 0
+    prediction_correct = 0
+    val_loss = 0
+    val_steps = 0
+    with torch.no_grad():
+        for idx, (inputs, target_control) in enumerate(val_loader):
+            prediction_total += inputs.size(0)
+
+            # Run inputs through model, save prediction
+            predicted_control = model(inputs)
+            # Round to nearest 100 MHz increment
+            predicted_control = np.round(predicted_control,1)
+            prediction_correct += (target_control == predicted_control).sum().item()
+
+            loss = loss_fn(predicted_control, target_control)
+            val_loss += loss.numpy()
+            val_steps += 1;
+
+
+    return val_loss/val_steps, prediction_correct/prediction_total
+
+def test(model, test_loader):
+    prediction_total = 0
+    prediction_correct = 0
+    prediction_min = 0
+    prediction_max = 0
+    with torch.no_grad():
+        for idx, (inputs, target_control) in enumerate(test_loader):
+            prediction_total += inputs.size(0)
+            # Run inputs through model, save prediction
+            predicted_control = model(inputs)
+            # Round to nearest 100 MHz increment
+            predicted_control = np.round(predicted_control,1)
+
+            # Generate stats
+            prediction_correct += (target_control == predicted_control).sum().item()
```

Comment:
Great question!  This is the final test against data that was not used in training and validation, and is only hit if --leave-app-out was specified.  We're rounding any frequency decisions to the nearest 100MHz to match common CPU controls which provides some leeway.  All that said, I'm in favor of passing back loss as proposed.  

In the training loop we are passing back this same accuracy metric and using it for model selection and early termination via ASHA.  We are also passing back loss using MSE there, but it is only used for reporting.  We could change lines 166-167 to as called out in the TODO:
```
            metric="accuracy", #TODO: we may want to switch to using metric="loss", mode="min"
            mode="max",
```

To have ASHA use the loss instead, but we should investigate the tradeoff there

Reply:
Is STEP_SIZE a universal limitation across all of our CPU frequency control interfaces (beyond just the ones we already have IOGroups for)?

The reason I ask is because I wonder if the 100 MHz rounding behavior belongs in a controller helper's code (e.g., an agent, the FrequencyGovernor, or similar) instead. Then this (or other frequency recommenders) could recommend the best frequency for a given decision and the governor or other class could decide what to do when the exact frequency can't be specified.

Reply:
I believe some client hardware uses 50MHz steps.  On the non GPU side we know that NVIDIA hardware supported by the NVMLIOGroup has 7 & 8 MHz steps, and these changes should be moved to the train_gpu_model-pytorch.py code eventually.
I've updated the code to use a single 'evaluate()' function that provides accuracy and loss, which may be used instead of accuracy.